### ISSUE TITLE
Incorrect type inference with is_array, array_is_list negation in condition

### ISSUE THREAD

**Issue by Riimu** [vimeo/psalm#7469]
It seem psalm incorrectly infers array type in some cases. The following kind of line produces an unexpected error in psalm:

```php
if (!is_array($value) || !array_is_list($value)) {
```

The produced error:

```
ERROR: TypeDoesNotContainType - 5:30 - Type array<array-key, mixed> for $value is always !list<mixed>
```

Reproduction here: https://psalm.dev/r/63f2cb0d8e

Not entirely sure about the root cause of this one (so may be duplicate of existing issue). It seems however, that inverse of the condition [does not produce an error](https://psalm.dev/r/99aba2a7ab).
